File: srcr/src/days_2025/day5.rs
pub fn solve_day_5 (input: &str) {

    let mut valid: usize = 0;

    let (list_ids, list_id) = clean_data(&input);
    let result_part_2 = part_2(&input);

    // part 1
    for id in list_id {
        for (min, max) in &list_ids {
            if id >= *min && id <= *max {
                valid +=1;
                break;
            }
        }
    }

    println!("Résultat : {}", valid);
    println!("Résultat 2 : {}", result_part_2);

}


fn part_2(input: &str) -> usize {
    let (list_ids, _) = clean_data(input);

    let mut intervals = list_ids;
    intervals.sort_by_key(|&(min, _)| min);

    let mut merged: Vec<(usize, usize)> = Vec::new();

    for (min, max) in intervals {
        if merged.is_empty() {
            merged.push((min, max));
        } else {
            let last_idx = merged.len() - 1;
            let (last_min, last_max) = merged[last_idx];

            if min <= last_max + 1 {
                merged[last_idx] = (last_min, last_max.max(max));
            } else {
                merged.push((min, max));
            }
        }
    }
    merged.iter().map(|(min, max)| max - min + 1).sum()
}

fn clean_data(input: &str)-> (Vec<(usize, usize)>, Vec<usize>) {

let group = input.split("\n\n").collect::<Vec<&str>>();

    let list_ids = group[0]
    .split("\n")
    .filter(|r| !r.trim().is_empty())
    .map(|r| {
        let (min, max) = r.split_once("-").unwrap();
        (min.trim().parse().unwrap(), max.trim().parse().unwrap())
    })
    .collect::<Vec<(usize, usize)>>();

    let list_id = group[1]
        .split("\n")
        .filter(|i| !i.trim().is_empty())
        .map(|i| i.trim().parse::<usize>().unwrap())
        .collect::<Vec<usize>>();

    (list_ids, list_id)
}

File: srcd/bin/2025/day5.dart
class Day5_2025 {
  List<(int, int)> listIds;
  List<int> listId;
  int resultat_1 = 0;
  int resultat_2 = 0;

  Day5_2025._({
    required this.listIds,
    required this.listId,
  });

  factory Day5_2025(String input) {
    final List<String> group = input.split("\n\n").toList();

    final List<(int, int)> listIds =
        group.first.split("\n").where((r) => r.trim().isNotEmpty).map((r) {
      final parts = r.split("-");
      final min = int.parse(parts[0].trim());
      final max = int.parse(parts[1].trim());
      return (min, max);
    }).toList();

    final List<int> listId = group[1]
        .split("\n")
        .where((i) => i.trim().isNotEmpty)
        .map((i) => int.parse(i.trim()))
        .toList();

    return Day5_2025._(listIds: listIds, listId: listId);
  }

  (int, int) getResult() {
    for (var id in listId) {
      for (var i = 0; i < listIds.length; i++) {
        final (min, max) = listIds[i];

        if (id >= min && id <= max) {
          resultat_1 += 1;
          break;
        }
      }
    }

    resultat_2 = part_2();

    return (resultat_1, resultat_2);
  }

  int part_2() {
    final intervals = listIds;
    intervals.sort((a, b) => a.$1.compareTo(b.$1));
    List<(int, int)> merged = [];

    for (var (min, max) in intervals) {
      if (merged.isEmpty) {
        merged.add((min, max));
      } else {
        var lastIdx = merged.length - 1;
        var (lastMin, lastMax) = merged[lastIdx];

        if (min <= lastMax + 1) {
          merged[lastIdx] = (lastMin, max > lastMax ? max : lastMax);
        } else {
          merged.add((min, max));
        }
      }
    }
    return merged.fold(
        0, (sum, interval) => sum + (interval.$2 - interval.$1 + 1));
  }
}
